src/System.Private.CoreLib/src/System/ValueType.cs

```diff
@@ -36,4 +42,161 @@ public override int GetHashCode()
+
+                    // Fetch the value of the field on both types
+                    object thisField = RuntimeImports.RhBoxAny(ref Unsafe.Add(ref thisRawData, fieldOffset), fieldType);
+                    object thatField = RuntimeImports.RhBoxAny(ref Unsafe.Add(ref thatRawData, fieldOffset), fieldType);
+
+                    // Compare the fields
+                    if (thisField == null)
+                    {
+                        if (thatField != null)
+                            return false;
+                    }
+                    else if (!thisField.Equals(thatField))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        public override int GetHashCode()
+        {
+            int hashCode = this.EETypePtr.GetHashCode();
+
+            int numFields = __GetFieldHelper(GetNumFields, out _);
+
+            if (numFields == UseFastHelper)
+            {
+                hashCode ^= FastGetValueTypeHashCodeHelper(this.EETypePtr, ref this.GetRawData());
+            }
+            else
+            {
+                hashCode ^= RegularGetValueTypeHashCode(this.EETypePtr, ref this.GetRawData(), numFields);
+            }
+
+            return hashCode;
+        }
+
+        private static int FastGetValueTypeHashCodeHelper(EETypePtr type, ref byte data)
+        {
+            // Sanity check - if there are GC references, we should not be hashing bytes
+            Debug.Assert(!type.HasPointers);
+
+            int size = (int)type.ValueTypeSize;
+            int hashCode = 0;
+
+            for (int i = 0; i < size / 4; i++)
+            {
+                hashCode ^= Unsafe.As<byte, int>(ref Unsafe.Add(ref data, i * 4));
+            }
+
+            return hashCode;
+        }
+
+        private int RegularGetValueTypeHashCode(EETypePtr type, ref byte data, int numFields)
+        {
+            int hashCode = 0;
+
+            // We only take the hashcode for the first non-null field. That's what the CLR does.
+            for (int i = 0; i < numFields; i++)
+            {
+                int fieldOffset = __GetFieldHelper(i, out EETypePtr fieldType);
+                ref byte fieldData = ref Unsafe.Add(ref data, fieldOffset);
+
+                Debug.Assert(!fieldType.IsPointer);
+
+                if (fieldType.CorElementType == RuntimeImports.RhCorElementType.ELEMENT_TYPE_R4)
+                {
+                    hashCode = Unsafe.Read<float>(ref fieldData).GetHashCode();
+                }
+                else if (fieldType.CorElementType == RuntimeImports.RhCorElementType.ELEMENT_TYPE_R8)
+                {
+                    hashCode = Unsafe.Read<double>(ref fieldData).GetHashCode();
+                }
+                else if (fieldType.IsPrimitive)
+                {
+                    hashCode = FastGetValueTypeHashCodeHelper(fieldType, ref fieldData);
```

Comment:
I think you've correctly ported this from CoreCLR, but it appears to be yet another bug (not every primitive has a hashcode equal to its value)

Reply:
Also, I've confirmed that structs smaller than 4 bytes all have the same hash code on CoreCLR because of this and the fact that FastGetValueTypeHashCodeHelper . 😦

Reply:
https://github.com/dotnet/coreclr/issues/16427